pkg/ipam/ippool.go

```diff
@@ -0,0 +1,203 @@
+limitations under the License.
+*/
+package ipam
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math"
+	"net"
+	"sync"
+
+	"github.com/emirpasic/gods/maps/treemap"
+)
+
+type IPPool struct {
+	cidr      string
+	network   *net.IPNet
+	size      int
+	available *treemap.Map // int IP is the key, string IP is the value
+	sync.RWMutex
+}
+
+func NewIPPool(cidr string) (*IPPool, error) {
+	_, network, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return nil, err
+	}
+
+	ones, totalbits := network.Mask.Size()
+	size := int(math.Exp2(float64(totalbits-ones))) - 2 // don't count net and broadcast
+	if size < 2 {
+		return nil, fmt.Errorf("invalid prefix for CIDR %q", cidr)
+	}
+
+	pool := &IPPool{
+		cidr:      cidr,
+		network:   network,
+		size:      size,
+		available: treemap.NewWithIntComparator(),
+	}
+
+	startingIP := ipToInt(pool.network.IP) + 1
+
+	for i := 0; i < pool.size; i++ {
+		intIP := startingIP + i
+		ip := intToIP(intIP).String()
+		pool.available.Put(intIP, ip)
+	}
+
+	// TODO: RecordAvailability(cidr, size)
+
+	return pool, nil
+}
+
+func ipToInt(ip net.IP) int {
+	intIP := ip
+	if len(ip) == 16 {
+		intIP = ip[12:16]
+	}
+
+	return int(binary.BigEndian.Uint32(intIP))
+}
+
+func intToIP(ip int) net.IP {
+	netIP := make(net.IP, 4)
+	binary.BigEndian.PutUint32(netIP, uint32(ip))
+
+	return netIP
+}
+
+func StringIPToInt(stringIP string) int {
+	return ipToInt(net.ParseIP(stringIP))
+}
+
+func (p *IPPool) allocateOne() ([]string, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	iter := p.available.Iterator()
+	if iter.Last() {
+		p.available.Remove(iter.Key())
+		// TODO: RecordAllocateGlobalIP(p.cidr)
+
+		return []string{iter.Value().(string)}, nil
+	}
+
+	return nil, errors.New("insufficient IPs available for allocation")
+}
+
+func (p *IPPool) Allocate(num int) ([]string, error) {
+	if num < 0 {
+		return nil, errors.New("the number to allocate cannot be negative")
+	}
+
+	if num == 0 {
+		return []string{}, nil
+	}
+
+	if num == 1 {
+		return p.allocateOne()
+	}
+
+	if p.available.Size() < num {
+		return nil, fmt.Errorf("insufficient IPs available (%d) to allocate %d", p.available.Size(), num)
+	}
+
+	retIPs := make([]string, num)
+
+	var prevIntIP, firstIntIP, current int
+
+	p.Lock()
+	defer p.Unlock()
+
+	iter := p.available.Iterator()
+	for iter.Next() {
+		intIP := iter.Key().(int)
+		retIPs[current] = iter.Value().(string)
+
+		if current == 0 || prevIntIP+1 != intIP {
+			firstIntIP = intIP
+			prevIntIP = intIP
+			retIPs[0] = retIPs[current]
+			current = 1
+
+			continue
+		}
+
+		prevIntIP = intIP
+		current++
+
+		if current == num {
+			for i := 0; i < num; i++ {
+				p.available.Remove(firstIntIP)
+				firstIntIP++
+			}
+
+			// TODO: RecordAllocateGlobalIPs(p.cidr, num)
+
+			return retIPs, nil
+		}
+	}
+
+	return nil, fmt.Errorf("unable to allocate a contiguous block of %d IPs - available pool size is %d",
+		num, p.available.Size())
+}
+
+func (p *IPPool) Release(ips ...string) {
```

Comment:
What if someone tries to release an IP which doesn't even belong to this CIDR? With available+allocated we had an implicit check. We should add a check to make sure we're not adding an IP outside of CIDR to the pool coz then it may get allocated by mistake. This can happen if consumer code is not checking change in CIDRs and cleaning old CIDR's IPs correctly.

Reply:
yeah I was thinking that too.

Reply:
Done.